---
title: "Kallisto_ processing"
output: 
  html_document:
    df_print: paged
author: "Sangeev"
date: "`r Sys.time()`"
---

<style type="text/css">
.main-container {
  max-width: 1500px;
  margin-left: auto;
  margin-right: auto;
}
</style>

```{r}
# Prevent printing of warnings and such in the HTML
knitr::opts_chunk$set(warning = FALSE, message = FALSE, fig.align = "center", eval = FALSE)
```

---
```{r}
library(clusterProfiler)
library(DESeq2) 
library(tidyverse)
library(topGO)
library(org.Mm.eg.db)
library(rtracklayer)
library(scales)
library(patchwork)
```

## Read in Kallisto output 
```{r}
file.loc <- list.files(path = "../alignment/kalisto/output/", 
                       pattern = "tsv",
                       full.names = TRUE,
                       recursive = TRUE)
file.loc

kallisto_data <- sapply(file.loc, read_tsv, simplify = FALSE) %>% # Read in multiple kallisto files
  bind_rows(.id = "sample") # Join them into a single tibble

kallisto_data <- kallisto_data %>%
  separate(sample, into = letters[1:6], sep = "/") %>% # Separate the file name to get sample name
  dplyr::select(f, target_id, est_counts, tpm, eff_length, length) # Keep only these 3 columns for subsequent analysis
```



```{r}
kallisto_clean <- kallisto_data %>% 
  separate(target_id, into = letters[1:2], sep = "\\|") %>%
  dplyr::rename("gene_id"="b") %>% 
  dplyr::rename("sample"="f") %>% 
  dplyr::rename("transcript"="a") 

kallisto_clean %>% 
  group_by(sample, gene_id) %>% 
  summarise(sum_tpm=sum(tpm)) %>%
  tally(sum_tpm)

filtered <- kallisto_clean %>% 
  group_by(sample, gene_id) %>% 
  summarise(sum_tpm=sum(tpm)) %>%
  pivot_wider(names_from = sample, values_from = sum_tpm) %>%
  rename("gene"="gene_id")

filtered
write_csv(filtered, "../data_frames/filtered.csv")
filtered <- read_csv("../data_frames/filtered.csv")
```


#Looks like mut1,mut3- wt1,wt2 should be used for differential expression analysis. 
```{r}
 
  correlation <- filtered %>% 
  dplyr::select(-gene)

corr <- correlation %>%
   filter(mut1< 5) %>%
   filter(wt1< 5) %>%
   filter(mut2 < 5) %>%
   filter(wt2 < 5) %>%
   filter(mut3 < 5) %>%
   filter(wt3 < 5) %>%
   filter(!is.na(mut1) & !is.infinite(mut1) & mut1 != 0) %>%
   filter(!is.na(wt1) & !is.infinite(wt1) & wt1 != 0) %>%
   filter(!is.na(mut2) & !is.infinite(mut2) & mut2 != 0) %>%
   filter(!is.na(wt2) & !is.infinite(wt2) & wt2 != 0) %>%
   filter(!is.na(mut3) & !is.infinite(mut3) & mut3 != 0) %>%
   filter(!is.na(wt3) & !is.infinite(wt3) & wt3 != 0) %>%
  rename(c("mut1"="KO1", "mut2"="KO2", "mut3"="KO3", "wt1"="CTRL1", "wt2"="CTRL2", "wt3"="CTRL3")) %>%
  select(CTRL1, CTRL2, CTRL3, KO1, KO2, KO3) %>%
  GGally::ggpairs(log2(df+1))

corr

ggsave(plot = last_plot(),filename = "/data2/sangeev/2022_organoid/figures/corr1.pdf", dpi = 600, width = 8, height =5)
```


```{r}
correlation2 <- correlation %>%
   rename(c("KO1"="mut1", "KO2"="mut2", "KO3"="mut3", "CTRL1"="wt1", "CTRL2"= "wt2", "CTRL3"="wt3")) %>%
  dplyr::select(CTRL1, CTRL2, CTRL3, KO1, KO2, KO3) 
  
corr_wt <- correlation2 %>%  dplyr::select(CTRL1, CTRL2, CTRL3)
corr_mut <- correlation2 %>%  dplyr::select(KO1, KO2, KO3) 

correlation2 <- correlation2[rowSums(correlation2)>9,] #keep genes with some counts
set.seed(123)
df <- correlation2[sample.int(nrow(correlation2), 1e3),] #random 1K genes

GGally::ggpairs(log10(df+1))
```

```{r}
GGscatterPlot <- function(data, mapping, ..., 
                        method = "pearson") {

#Get correlation coefficient
    x <- GGally::eval_data_col(data, mapping$x)
    y <- GGally::eval_data_col(data, mapping$y)

    cor <- cor(x, y, method = method)
#Assemble data frame
    df <- data.frame(x = x, y = y)
# PCA
    nonNull <- x!=0 & y!=0
    dfpc <- prcomp(~x+y, df[nonNull,])
    df$cols <- predict(dfpc, df)[,1]
# Define the direction of color range based on PC1 orientation:
    dfsum <- x+y
    colDirection <- ifelse(dfsum[which.max(df$cols)] < 
                               dfsum[which.min(df$cols)],
                           1,
                           -1)
#Get 2D density for alpha
    dens2D <- MASS::kde2d(df$x, df$y)
    df$density <- fields::interp.surface(dens2D , 
                                         df[,c("x", "y")])

if (any(df$density==0)) {
    mini2D = min(df$density[df$density!=0]) #smallest non zero value
    df$density[df$density==0] <- mini2D
}
#Prepare plot
    pp <- ggplot(df, aes(x=x, y=y, color = cols, alpha = 1/density)) +
                ggplot2::geom_point(shape=16, show.legend = FALSE) +
                ggplot2::scale_color_viridis_c(direction = colDirection) +
#                scale_color_gradient(low = "#0091ff", high = "#f0650e") +
                ggplot2::scale_alpha(range = c(.05, .6)) +
                ggplot2::geom_abline(intercept = 0, slope = 1, col="darkred") +
                ggplot2::geom_label(
                        data = data.frame(
                                        xlabel = min(x, na.rm = TRUE),
                                        ylabel = max(y, na.rm = TRUE),
                                        lab = round(cor, digits = 3)),
                        mapping = ggplot2::aes(x = xlabel, 
                                               y = ylabel, 
                                               label = lab),
                        hjust = 0, vjust = 1,
                        size = 3, fontface = "bold",
                        inherit.aes = FALSE # do not inherit anything from the ...
                        ) +
                theme_minimal()

return(pp)
}
```


```{r}
library(GGally)

correlation3 <- corr_wt[rowSums(corr_wt)>5,] #keep genes with some counts
set.seed(123)
df <- correlation3[sample.int(nrow(correlation3), 5000),] #random 1K genes

GGally::ggpairs(log10(df+3),
                lower = list(continuous = GGscatterPlot),
                upper = list(continuous = wrap("cor", method= "pearson")))
ggsave(plot = last_plot(),filename = "/data2/sangeev/2022_organoid/figures/wt_corr.pdf", dpi = 600, width = 4, height =4)
```

```{r}
correlation4 <- corr_mut[rowSums(corr_mut)>5,] #keep genes with some counts
set.seed(123)
df2 <- correlation4[sample.int(nrow(correlation4), 5000),] #random 1K genes

GGally::ggpairs(log10(df2+3),
                lower = list(continuous = GGscatterPlot),
                upper = list(continuous = wrap("cor", method= "pearson")))
ggsave(plot = last_plot(),filename = "/data2/sangeev/2022_organoid/figures/mut_corr.pdf", dpi = 600, width = 4, height =4)
```



```{r}
#the correlation between CTRL1,2, and KO 4,6 are perfect). but not good with wt 3 and mut 2. 
GGally::ggduo(df,
              c(1,3), 
              c(1,2),
              types = list(continuous = GGscatterPlot)) +
    theme_bw() +
  scale_x_continuous(trans = "log10",
                     limits = c(1e0, 1e5)) +
   scale_y_continuous(trans = "log10",
                      limits = c(1e0, 1e5)) +
    theme(panel.grid.minor = element_blank())
ggsave(plot = last_plot(),filename = "/data2/sangeev/2022_organoid/figures/corr4.pdf", dpi = 600, width = 5, height =5)
```


```{r}
# Install the corrplot package if not already installed
install.packages("corrplot")

# Load the corrplot package
library(corrplot)

correlation2 <- correlation %>%
   rename(c("KO1"="mut1", "KO2"="mut2", "KO3"="mut3", "CTRL1"="wt1", "CTRL2"= "wt2", "CTRL3"="wt3")) %>%
  dplyr::select(CTRL1, CTRL2, CTRL3, KO1, KO2, KO3) 
  
corr_wt <- correlation2 %>%  dplyr::select(CTRL1, CTRL2, CTRL3)
corr_mut <- correlation2 %>%  dplyr::select(KO1, KO2, KO3) 

# Create a correlation matrix
cor_matrix <- cor(correlation2 %>%
                    filter(!is.na(KO1) & !is.infinite(KO1) & KO1 != 0) %>%
   filter(!is.na(CTRL1) & !is.infinite(CTRL1) & CTRL1 != 0) %>%
   filter(!is.na(KO2) & !is.infinite(KO2) & KO2 != 0) %>%
   filter(!is.na(CTRL2) & !is.infinite(CTRL2) & CTRL2 != 0) %>%
   filter(!is.na(KO3) & !is.infinite(KO3) & KO3 != 0) %>%
   filter(!is.na(CTRL3) & !is.infinite(CTRL3) & CTRL3 != 0)) 

# Create a correlation matrix heatmap
corrplot(cor_matrix, method = "color")

ggsave(corrplot(cor_matrix, method = "color"),filename = "/data2/sangeev/2022_organoid/figures/corr2.pdf", dpi = 600, width = 5, height =5))
```

```{r}
# Install the FactoMineR package if not already installed
install.packages("FactoMineR")

# Load the FactoMineR package
library(FactoMineR)
library(factoextra) 

# Create a correlation circle plot
pca_res <- PCA(correlation, graph = FALSE)
fviz_pca_var(pca_res, col.var = "contrib", col.ind = "cos2")
```


```{r fig.width=8, fig.height=6}
mut1 <- corr_mut %>%
  filter(!is.na(KO1) & !is.infinite(KO1) & KO1 != 0) %>%
   filter(!is.na(KO2) & !is.infinite(KO2) & KO2 != 0) %>%
   ggplot(., aes(x = KO1, y = KO2)) +
  geom_point(size = 1, alpha = 0.5, col = "gray") +
  ggpointdensity::geom_pointdensity() +
  scale_x_continuous(trans = "log10", limits= c(1e-2,1e5))+
  scale_y_continuous(trans = "log10", limits= c(1e-2,1e5))+
  guides(col = "none") +
  ggpubr::theme_pubr() +
  coord_equal() +
  scale_color_viridis_c(option = "B",
                         name = "Neighboring\npoints",
                         guide = FALSE)+
  ggpubr::stat_cor(aes(label = ..r.label..)) +
  #geom_smooth(method = "lm") + 
  labs(x = "KO1", y = "KO2")

mut2 <- corr_mut %>%
  filter(!is.na(KO2) & !is.infinite(KO2) & KO2 != 0) %>%
   filter(!is.na(KO3) & !is.infinite(KO3) & KO3 != 0) %>%
   ggplot(., aes(x = KO2, y = KO3)) +
  geom_point(size = 1, alpha = 0.5, col = "gray") +
  ggpointdensity::geom_pointdensity() +
  scale_x_continuous(trans = "log10", limits= c(1e-2,1e5))+
  scale_y_continuous(trans = "log10", limits= c(1e-2,1e5))+
  guides(col = "none") +
  ggpubr::theme_pubr() +
  coord_equal() +
  scale_color_viridis_c(option = "B",
                         name = "Neighboring\npoints",
                         guide = FALSE)+
  ggpubr::stat_cor(aes(label = ..r.label..)) +
  #geom_smooth(method = "lm") + 
  labs(x = "KO2", y = "KO3")

mut3 <- corr_mut %>%
  filter(!is.na(KO1) & !is.infinite(KO1) & KO1 != 0) %>%
   filter(!is.na(KO3) & !is.infinite(KO3) & KO3 != 0) %>%
   ggplot(., aes(x = KO1, y = KO3)) +
  geom_point(size = 1, alpha = 0.5, col = "gray") +
  ggpointdensity::geom_pointdensity() +
  scale_x_continuous(trans = "log10", limits= c(1e-2,1e5))+
  scale_y_continuous(trans = "log10", limits= c(1e-2,1e5))+
  guides(col = "none") +
  ggpubr::theme_pubr() +
  coord_equal() +
  scale_color_viridis_c(option = "B",
                         name = "Neighboring\npoints",
                         guide = FALSE)+
  ggpubr::stat_cor(aes(label = ..r.label..)) +
  #geom_smooth(method = "lm") + 
  labs(x = "KO1", y = "KO3")

mut1

mut1|mut2|mut3
```

```{r}
wt1 <- corr_wt %>%
  filter(!is.na(CTRL1) & !is.infinite(CTRL1) & CTRL1 != 0) %>%
   filter(!is.na(CTRL2) & !is.infinite(CTRL2) & CTRL2 != 0) %>%
   ggplot(., aes(x = CTRL1, y = CTRL2)) +
  geom_point(size = 1, alpha = 0.5, col = "gray") +
  ggpointdensity::geom_pointdensity() +
  scale_x_continuous(trans = "log10", limits= c(1e-2,1e5))+
  scale_y_continuous(trans = "log10", limits= c(1e-2,1e5))+
  guides(col = "none") +
  ggpubr::theme_pubr() +
  coord_equal() +
  scale_color_viridis_c(option = "B",
                         name = "Neighboring\npoints",
                         guide = FALSE)+
  ggpubr::stat_cor(aes(label = ..r.label..)) +
  #geom_smooth(method = "lm") + 
  labs(x = "CTRL1", y = "CTRL2")

wt2 <- corr_wt %>%
  filter(!is.na(CTRL3) & !is.infinite(CTRL3) & CTRL3 != 0) %>%
   filter(!is.na(CTRL2) & !is.infinite(CTRL2) & CTRL2 != 0) %>%
   ggplot(., aes(x = CTRL3, y = CTRL2)) +
  geom_point(size = 1, alpha = 0.5, col = "gray") +
  ggpointdensity::geom_pointdensity() +
  scale_x_continuous(trans = "log10", limits= c(1e-2,1e5))+
  scale_y_continuous(trans = "log10", limits= c(1e-2,1e5))+
  guides(col = "none") +
  ggpubr::theme_pubr() +
  coord_equal() +
  scale_color_viridis_c(option = "B",
                         name = "Neighboring\npoints",
                         guide = FALSE)+
  ggpubr::stat_cor(aes(label = ..r.label..)) +
  #geom_smooth(method = "lm") + 
  labs(x = "CTRL3", y = "CTRL2")

wt3 <- corr_wt %>%
  filter(!is.na(CTRL1) & !is.infinite(CTRL1) & CTRL1 != 0) %>%
   filter(!is.na(CTRL3) & !is.infinite(CTRL3) & CTRL3 != 0) %>%
   ggplot(., aes(x = CTRL1, y = CTRL3)) +
  geom_point(size = 1, alpha = 0.5, col = "gray") +
  ggpointdensity::geom_pointdensity() +
  scale_x_continuous(trans = "log10", limits= c(1e-2,1e5))+
  scale_y_continuous(trans = "log10", limits= c(1e-2,1e5))+
  guides(col = "none") +
  ggpubr::theme_pubr() +
  coord_equal() +
  scale_color_viridis_c(option = "B",
                         name = "Neighboring\npoints",
                         guide = FALSE)+
  ggpubr::stat_cor(aes(label = ..r.label..)) +
  #geom_smooth(method = "lm") + 
  labs(x = "CTRL1", y = "CTRL3")
```


### Executing the DESeq2 DEG analysis, GO analysis across all three ontologies, and KEGG analysis
#### Read in appris
```{r}
appris <- read_delim("/data/riboseq/Supp/seqs/mouse/vm14_current/mouse_appris_data.principal.txt", "\t", col_names = F)
names(appris)[1:3] <- c("Gene", "ENSEMBL", "Transcript") 
```


## Prepare data for differential gene expression analysis
```{r}
filtered2 <- filtered %>% 
  mutate(mut1=round(mut1),
    mut3= round(mut3), 
    wt1= round(wt1), 
    wt2=round(wt2)) %>%
  dplyr::select(gene, mut1, mut3, wt1, wt2)

#need matrix for the packages. 
counts_matrix <- filtered2[, 2:5] %>%
  as.matrix() 
  counts_matrix
  
  rownames(counts_matrix) <- filtered$gene
  counts_matrix
  
  #create a new dataframe with information about the data
  
  condition_df <- data.frame(conditions= factor(rep(c("Mutant", "Wild-Type"), each=2),
                                                levels = c("Wild-Type", "Mutant")))
  
  #when you have batch effects
#   condition_df <- data.frame(conditions= factor(rep(c("Mutant", "Wild-Type"), each=3),
#                                                levels = c("Wild-Type", "Mutant")),
#                             batch=c(1,1,2,1,1,2)) #you need this to show how the mutant changes in respect to the WT...so the levels of mutant comes after wt (reference genome). when you just set factors, it will set based on alphabetical order- mutant first thats why you need levels. 

condition_df

head(counts_matrix)

```
## Perform differential gene expression analysis
```{r}
# When you have batch effects you need to account for:-
# data_for_deseq <- DESeqDataSetFromMatrix(countData= counts_matrix,
 #                                        colData= condition_df,
 #                                        design= ~batch+conditions)

data_for_deseq <- DESeqDataSetFromMatrix(countData= counts_matrix,
                                         colData= condition_df,
                                         design= ~conditions)
data_for_deseq

diff_exp_analysis <- DESeq(data_for_deseq)
diff_exp_analysis

diff_exp_results <- results(diff_exp_analysis)
diff_exp_results
```
#PCA plot- just to confirm how separated/close the samples are. looks like mut2,mut3 and wt1,wt2 were the right choices.
```{r}
test <- varianceStabilizingTransformation(data_for_deseq, blind = F)
plotPCA(test, intgroup=c("conditions")) + geom_text(aes(label = colnames(test)), color = "black")
```


```{r}
plotMA(diff_exp_results)

# Log foldchange shrinkage
diff_exp_results_shrink <- lfcShrink(diff_exp_analysis, 
                                     coef = 2,
                                     type = "apeglm")

plotMA(diff_exp_results)
plotMA(diff_exp_results_shrink)
```
## Identify differentially expressed genes
```{r}
diff_exp_tibble <- diff_exp_results_shrink%>%
  as.tibble() %>%
  mutate(gene= rownames(diff_exp_results_shrink)) %>%
  dplyr::select(gene, everything())

#dwn regulated

diff_exp_tibble %>%
  filter(log2FoldChange < 0, padj <0.05) #why not use p value instead?

#up-regulated
diff_exp_tibble %>%
  filter(log2FoldChange > 0, padj <0.05) %>% arrange(desc(log2FoldChange))

```

```{r}
#volcano plots
diff_exp_tibble %>%
  ggplot(., aes(x=log2FoldChange, y= -log10(padj))) +
  geom_point()

new_diff_exp_tibble <- diff_exp_tibble %>%
  mutate(new_y = -log10(padj),
         new_padj = ifelse(new_y > 5, 5, new_y),
         color= new_padj > -log10(0.05) & abs(log2FoldChange)>2)


new_diff_exp_tibble%>%
  ggplot(., aes(x=log2FoldChange, y= new_padj, col=color)) + #significant and large change...
  geom_point() 

#blue=significant and have at lst 2 fold change gene expression, right-up, left-down
```
#### Read in appris
```{r}
appris <- read_gff("/data/riboseq/Supp/seqs/mouse/vm14_current/mouse_appris_data.principal.txt", "\t", col_names = F)
appris 
data <- diff_exp_tibble %>%  separate(gene, into = LETTERS[1:2], sep = "\\.") %>%
  rename("ENSEMBL"= "A") 
tidied <- data %>%
left_join(., appris, by="ENSEMBL") %>%
  filter(!is.na(Gene))

tidied
tmp <- bitr(tidied$ENSEMBL, fromType = "ENSEMBL", toType = c("ENTREZID", "SYMBOL"), OrgDb = "org.Mm.eg.db")
tidied2 <- left_join(tidied, tmp, "ENSEMBL") %>% filter(!is.na(ENTREZID)) 

tidied2 
```

## Perform functional analyses using clusterProfiler
http://yulab-smu.top/clusterProfiler-book/

```{r}

#enrichment analysis- genes that have significantly higher expression in the mutant. 
sig_up_genes <- tidied2 %>%
  filter(log2FoldChange > 0, padj <0.05) %>% #no result if i use padj of even 0.5
  pull(ENTREZID) 

sig_up_genes

#Kegg analysis- overrepresentation analysis....
sig_up_genes_kegg_enrich <- enrichKEGG(gene = sig_up_genes,
                                       organism = "mmu") 
sig_up_genes_kegg_enrich 
barplot(sig_up_genes_kegg_enrich)
dotplot(sig_up_genes_kegg_enrich) 

sig_dwn_genes <- tidied2 %>%
  filter(log2FoldChange < 0,  padj <0.05) %>%
  pull(ENTREZID) 

sig_dwn_genes

sig_dwn_genes_kegg_enrich <- enrichKEGG(gene = sig_dwn_genes,
                                       organism = "mmu") 
sig_dwn_genes_kegg_enrich

barplot(sig_dwn_genes_kegg_enrich)
dotplot(sig_dwn_genes_kegg_enrich)

cnetplot(sig_up_genes_kegg_enrich)
cnetplot(sig_dwn_genes_kegg_enrich, layout = "circle") #genes that are involved in the specific pathways..can play around with the layouts...

dotplot(sig_up_genes_kegg_enrich)
dotplot(sig_dwn_genes_kegg_enrich)
```

```{r}
#Gene Set enrichment analysis- work with actaul fold gene conversion, 

tidied3 <- tidied2 %>%
  filter(!is.na(log2FoldChange)) %>% #remove genes with NA in fold changes
  mutate(FoldChange= 2^log2FoldChange) %>% #actual fold changed
arrange(desc(FoldChange))

genes_With_fold_changes <- tidied3$FoldChange
names(genes_With_fold_changes) <- tidied3$ENTREZID

head(genes_With_fold_changes)


#GSEA- identify small but consistent changes in funtional changes

kegg_gsea <- gseKEGG(geneList = genes_With_fold_changes,
                     organism = "mmu")
kegg_gsea

dotplot(kegg_gsea)

#can try upset plot in new R package

heatplot(kegg_gsea) #
heatplot(kegg_gsea, foldChange = genes_With_fold_changes)


```

